sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/identifiers.scala

```diff
@@ -17,16 +17,22 @@
 
 package org.apache.spark.sql.catalyst
 
+import org.apache.spark.sql.connector.catalog.CatalogManager.SESSION_CATALOG_NAME
+import org.apache.spark.sql.internal.{SQLConf, StaticSQLConf}
+
 /**
  * An identifier that optionally specifies a database.
  *
  * Format (unquoted): "name" or "db.name"
  * Format (quoted): "`name`" or "`db`.`name`"
  */
-sealed trait IdentifierWithDatabase {
+sealed trait CatalystIdentifier {
+  private var _catalog: Option[String] = None
```

Comment:
This is a bit ugly... I think we should add `catalog: Option[String]` to `TableIdentifier` and `FunctionIdentifier` directly, similar to how `database` is handled.

Reply:
moved catalog to constructor